src/app/server/ArlStorage.h

```diff
@@ -0,0 +1,148 @@
+
+    public:
+        DecodableEntry() = default;
+
+        /**
+         * Reader decodes into a staging entry, which is then unstaged
+         * into a member entry.
+         */
+        CHIP_ERROR Decode(TLV::TLVReader & reader);
+
+        Entry & GetEntry() { return mEntry; }
+
+        const Entry & GetEntry() const { return mEntry; }
+
+    public:
+        static constexpr bool kIsFabricScoped = true;
+
+        void SetFabricIndex(FabricIndex fabricIndex) { mEntry.fabricIndex = fabricIndex; }
+
+    private:
+        Entry mEntry;
+
+        StagingEntry mStagingEntry;
+    };
+
+    /**
+     * Used for encoding access restriction entries.
+     *
+     * Typically used temporarily on the stack to encode:
+     * - source: system level access restriction entry
+     * - staging: generated cluster level code
+     * - destination: TLV
+     */
+    class EncodableEntry
+    {
+        using Entry              = Access::AccessRestriction::Entry;
+        using StagingEntry       = Clusters::AccessControl::Structs::AccessRestrictionEntryStruct::Type;
+        using StagingRestriction = Clusters::AccessControl::Structs::AccessRestrictionStruct::Type;
+
+    public:
+        EncodableEntry(std::shared_ptr<Entry> entry) : mEntry(entry) {}
+
+        /**
+         * Constructor-provided entry is staged into a staging entry,
+         * which is then encoded into a writer.
+         */
+        CHIP_ERROR EncodeForRead(TLV::TLVWriter & writer, TLV::Tag tag, FabricIndex fabric) const;
+
+        /**
+         * Constructor-provided entry is staged into a staging entry,
+         * which is then encoded into a writer.
+         */
+        CHIP_ERROR EncodeForWrite(TLV::TLVWriter & writer, TLV::Tag tag) const;
+
+        /**
+         * Constructor-provided entry is staged into a staging entry.
+         */
+        CHIP_ERROR Stage() const;
+
+        StagingEntry & GetStagingEntry() { return mStagingEntry; }
+
+        const StagingEntry & GetStagingEntry() const { return mStagingEntry; }
+
+    public:
+        static constexpr bool kIsFabricScoped = true;
+
+        FabricIndex GetFabricIndex() const { return mEntry->fabricIndex; }
+
+    private:
+        std::shared_ptr<Entry> mEntry;
+
+        mutable StagingEntry mStagingEntry;
+        mutable StagingRestriction mStagingRestrictions[CHIP_CONFIG_ACCESS_RESTRICTION_MAX_RESTRICTIONS_PER_ENTRY];
+    };
+
+    virtual ~ArlStorage() = default;
+
+    /**
+     * Initialize should be called after chip::Access::AccessControl is initialized.
+     *
+     * Implementations should take this opportunity to populate AccessControl with ARL entries
+     * loaded from persistent storage. A half-open range of fabrics [first, last) is provided
+     * so this can be done on a per-fabric basis.
+     *
+     * Implementations should also install an entry change listener on AccessControl to maintain
+     * ARL entries in persistent storage as they are changed.
```

Comment:
How is this done or supposed to be done.

I would have expected `ArlStorage` interface to have accessors for the ARL entries.

Reply:
This should be more clear now.  AccessRestrictionProvider is the single source-of-truth for the data.